packages/utils/src/index.ts

```diff
@@ -12,27 +12,32 @@ export const escapeRegExp = (str: string) =>
   (str || '').replace(/[\-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|]/g, '\\$&');
 
 // { key: value, ... } => { key: mapFn(value, key), ... }
-export const mapKeys = <T, U>(
-  obj: Record<string, T>,
-  func: (value: T, key: string, obj: Record<string, T>) => U
-): Record<string, U> =>
-  Object.entries(obj).reduce((acc, [key, value]) => ({ ...acc, [key]: func(value, key, obj) }), {});
+export const mapKeys = <T, K extends keyof T, R>(
+  obj: T,
+  func: (value: T[K], key: K, obj: T) => R
+) =>
+  Object.entries(obj).reduce(
+    (acc, [key, value]) => ({ ...acc, [key]: func(value, key as K, obj) }),
+    {} as Record<K, R>
+  );
 
 // { key: value, ... } => { mapFn(key, value): value, ... }
-export const mapKeyNames = <T>(
-  obj: Record<string, T>,
-  func: (key: string, value: T, obj: Record<string, T>) => string
-): Record<string, T> =>
+export const mapKeyNames = <T, K extends keyof T, R extends string>(
+  obj: T,
+  func: (key: K, value: T[K], obj: T) => R
+) =>
   Object.entries(obj).reduce(
-    (acc, [key, value]) => ({ ...acc, [func(key, value, obj)]: value }),
-    {}
+    (acc, [key, value]) => ({ ...acc, [func(key as K, value, obj)]: value }),
+    {} as Record<R, T[K]>
   );
 
-export const resolveAllKeys = async <T>(obj: Record<string, Promise<T>>) => {
-  const returnValue: Record<string, T> = {};
-  const errors: Record<string, unknown> = {};
+export const resolveAllKeys = async <V, T extends Record<string, Promise<V>>, K extends keyof T>(
+  obj: T
+) => {
+  const returnValue = {} as Record<K, V>;
+  const errors = {} as Record<K, unknown>;
```

Comment:
This will make it correctly type inputs like `{thing: Promise.resolve(1),another: Promise.resolve("something")}` as `Promise<{thing:number,another: string}>` rather than `Record<"thing" | "another", number | string>`

```suggestion
type Await<TPromise extends Promise<any>> = TPromise extends Promise<infer Value> ? Value : never;

export const resolveAllKeys = async <T extends Record<string, Promise<any>>>(obj: T) => {
  const returnValue = {} as { [Key in keyof T]: Await<T[Key]> };
  const errors = {} as Record<keyof T, unknown>;
```